core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala

```diff
@@ -53,35 +55,133 @@ case class ConnectionParams(
+  private val ConnectionTimeoutParamPath = "connection-timeout"
+  private val ConnectionTimeoutParamName = "connection_timeout"
+  private val HeartbeatParamName = "heartbeat"
+  private val ChannelMaxParamName = "channel_max"
+  private val AuthMechanismParamName = "auth_mechanism"
+  sealed case class UriQueryParams(
+                                    connectionTimeout: Int = DefaultConnectionTimeout,
+                                    heartbeat: Int = ConnectionFactory.DEFAULT_HEARTBEAT,
+                                    channelMax: Int = ConnectionFactory.DEFAULT_CHANNEL_MAX,
+                                    authMechanism: DefaultSaslConfig = DefaultSaslConfig.PLAIN
+                                  )
+
+  def fromConfig(config: Config = RabbitConfig.connectionConfig): ConnectionParams = {
+    if (config.hasPath(UriPramPath)) fromUri(new URI(config.getString(UriPramPath))) else fromParameters(config)
+  }
+
+  @deprecated(message = "The parameters configuration is deprecated if favor of the URL configuration", since = "2.1.0")
+  private def fromParameters(config: Config): ConnectionParams = {
+    val hosts = readHosts(config).toArray
+    val port = config.getInt(PortPramPath)
+
+    ConnectionParams(
+      hosts = hosts.map { h => new Address(h, port) },
+      connectionTimeout = config.getDuration(ConnectionTimeoutParamPath, java.util.concurrent.TimeUnit.MILLISECONDS).toInt,
+      username = config.getString(UsernamePramPath),
+      password = config.getString(PasswordPramPath),
+      virtualHost = config.getString(VirtualHostPramPath),
+      ssl = config.getBoolean(SslPramPath)
+    )
+  }
+
+  private def fromUri(uri: URI): ConnectionParams = {
+    val uriAuthority = uri.getAuthority
+    val (hosts, (username, password)) = uriAuthority.indexOf('@') match {
+      case -1 =>
+        parseUriHosts(uriAuthority) -> parseUserPassword(None)
+      case idx =>
+        parseUriHosts(uriAuthority.substring(idx + 1)) -> parseUserPassword(Some(uriAuthority.substring(0, idx)))
+    }
+    val params = parseAndValidateUriQuery(Option(uri.getQuery))
+    ConnectionParams(
+      hosts = hosts,
+      username = username,
+      password = password,
+      virtualHost = Option(uri.getPath).fold(ConnectionFactory.DEFAULT_VHOST)(_.substring(1)),
+      ssl = TlsProtectedScheme == uri.getScheme,
+      connectionTimeout = params.connectionTimeout,
+      requestedHeartbeat = params.heartbeat,
+      requestedChannelMax = params.channelMax,
+      saslConfig = params.authMechanism
+    )
+  }
+
+  private def parseUriHosts(hosts: String): Seq[Address] = {
+    hosts.split(",").map(Address.parseAddress)
+  }
+
+  private def parseUserPassword(auth: Option[String]): (String, String) = {
+    val default = (ConnectionFactory.DEFAULT_USER, ConnectionFactory.DEFAULT_PASS)
+    auth.fold(default) { candidate =>
+      candidate.split(":").toList match {
+        case usr :: psswd :: _ => (usr, psswd)
+        case _ => default
```

Comment:
Why don't we fail here? I'd like to know right away in this case

Reply:
Good catch